pkg/executor/join/hash_join_spill_helper.go

```diff
@@ -0,0 +1,607 @@
+	if h.buildRowsInDisk[workerID] == nil {
+		h.buildRowsInDisk[workerID] = make([]*chunk.DataInDiskByChunks, h.hashJoinExec.partitionNumber)
+		h.probeRowsInDisk[workerID] = make([]*chunk.DataInDiskByChunks, h.hashJoinExec.partitionNumber)
+	}
+
+	if h.buildRowsInDisk[workerID][partID] == nil {
+		inDisk := chunk.NewDataInDiskByChunks(h.buildSpillChkFieldTypes)
+		inDisk.GetDiskTracker().AttachTo(h.diskTracker)
+		h.buildRowsInDisk[workerID][partID] = inDisk
+
+		inDisk = chunk.NewDataInDiskByChunks(h.probeFieldTypes)
+		inDisk.GetDiskTracker().AttachTo(h.diskTracker)
+		h.probeRowsInDisk[workerID][partID] = inDisk
+	}
+
+	return h.spillSegmentsToDiskImpl(workerID, h.buildRowsInDisk[workerID][partID], segments)
+}
+
+func (h *hashJoinSpillHelper) spillSegmentsToDiskImpl(workerID int, disk *chunk.DataInDiskByChunks, segments []*rowTableSegment) error {
+	if cap(h.validJoinKeysBuffer[workerID]) == 0 {
+		h.validJoinKeysBuffer[workerID] = make([]byte, 0, maxRowTableSegmentSize)
+	}
+	h.validJoinKeysBuffer[workerID] = h.validJoinKeysBuffer[workerID][:0]
+	h.tmpSpillBuildSideChunks[workerID].Reset()
+
+	// Get row bytes from segment and spill them
+	for _, seg := range segments {
+		h.validJoinKeysBuffer[workerID] = h.generateSpilledValidJoinKey(seg, h.validJoinKeysBuffer[workerID])
+		rows := seg.getRowsBytesForSpill()
+
+		for i, row := range rows {
+			if h.tmpSpillBuildSideChunks[workerID].IsFull() {
+				err := disk.Add(h.tmpSpillBuildSideChunks[workerID])
+				if err != nil {
+					return err
+				}
+				h.tmpSpillBuildSideChunks[workerID].Reset()
+
+				err = triggerIntest(2)
+				if err != nil {
+					return err
+				}
+			}
+
+			h.tmpSpillBuildSideChunks[workerID].AppendInt64(0, int64(seg.hashValues[i]))
+			h.tmpSpillBuildSideChunks[workerID].AppendBytes(1, h.validJoinKeysBuffer[workerID][i:i+1])
+			h.tmpSpillBuildSideChunks[workerID].AppendBytes(2, row)
+		}
+	}
+
+	// Spill remaining rows in tmpSpillChunk[0]
+	if h.tmpSpillBuildSideChunks[workerID].NumRows() > 0 {
+		err := disk.Add(h.tmpSpillBuildSideChunks[workerID])
+		if err != nil {
+			return err
+		}
+		h.tmpSpillBuildSideChunks[workerID].Reset()
+	}
+	return nil
+}
+
+func (h *hashJoinSpillHelper) spillProbeChk(workerID int, partID int, chk *chunk.Chunk) error {
+	return h.probeRowsInDisk[workerID][partID].Add(chk)
+}
+
+func (h *hashJoinSpillHelper) init() {
+	if h.buildRowsInDisk == nil {
+		// It's the first time that spill is triggered
+		h.initTmpSpillBuildSideChunks()
+
+		h.buildRowsInDisk = make([][]*chunk.DataInDiskByChunks, h.hashJoinExec.Concurrency)
+		h.probeRowsInDisk = make([][]*chunk.DataInDiskByChunks, h.hashJoinExec.Concurrency)
+	}
+}
+
+func (h *hashJoinSpillHelper) spillRowTableImpl(partitionsNeedSpill []int, totalReleasedMemory int64) error {
+	workerNum := len(h.hashJoinExec.BuildWorkers)
+	errChannel := make(chan error, workerNum)
+
+	waiter := &sync.WaitGroup{}
```

Comment:
> util.WaitGroupWrapper

`util.WaitGroupWrapper` adds counter one by one, this will cause unexpected behaviour. We need to increase counter only once.

Reply:
> util.WaitGroupWrapper

done